### PR TITLE
Add #[repr(C)] for more future-proof byte mangling

### PR DESCRIPTION
**Pull request by ryoqun**
#### Problem

the setup part of the test introduced at https://github.com/solana-labs/solana/pull/7415#discussion_r1311294502 is fragile against rustc's abi-related changes.

For example, hashing change is needed due to updating rustc to 1.72: https://github.com/solana-labs/solana/pull/32961#discussion_r1311296564 

#### Summary of Changes

Use `repr(C)` to make test byte pattern generation more resilient to any abi changes.

i confirmed locally the updated hash in this pr is emitted now with both rust 1.69 and 1.72.

also some comment clarification is done for future me and others. :)

### CODE REVIEW

**Comment by ryoqun** [comment]
@t-nelson once this pr lands and #32961 is rebased, this code change should go away: https://github.com/solana-labs/solana/pull/32961#discussion_r1311296564

**Review by Lichtso — APPROVED** [review]

File: accounts-db/src/accounts_db.rs
Comment (Lichtso) [inline accounts-db/src/accounts_db.rs:12590]: Why the rewording of "cast" to "slice off"? This is still transmuting the entire thing, not just a slice of it, right?

File: accounts-db/src/accounts_db.rs
Comment (Lichtso) [inline accounts-db/src/accounts_db.rs:12594]: Don't think this is the right place for this comment to be seen. Better place it at the definition site of `AppendVecStoredAccountMeta`.

**Review by Lichtso — APPROVED** [review]

File: accounts-db/src/accounts_db.rs
Comment (ryoqun) [reply accounts-db/src/accounts_db.rs]: no particular reason other then i was thinking i've improve my english writing skill over the 3 years since I wrote the original comment. :) seems confusing; just reverted: 99ec8d53fc2

File: accounts-db/src/accounts_db.rs
Comment (ryoqun) [reply accounts-db/src/accounts_db.rs]: not so found of placing these stale-fated comments at definition if there's a unit test covering the specific future unintentional modifications.

Rather, I wanted to write what to do if this test failed in the future at this place. reworded a bit: 99ec8d53fc2

File: accounts-db/src/accounts_db.rs
Comment (ryoqun) [inline accounts-db/src/accounts_db.rs:12573]: I'm a bit surprised that plain tuple and tuple struct's abi differ; but i guess this is a life. ;)

**Comment by ryoqun** [comment]
i'm aware of competing prs (https://github.com/solana-labs/solana/pull/33082 and https://github.com/solana-labs/solana/pull/33083), but will just merge this to unblock https://github.com/solana-labs/solana/pull/32961

File: accounts-db/src/accounts_db.rs
Comment (apfitzge) [reply accounts-db/src/accounts_db.rs:12573]: In case you were curious about the the layout change for this specific tuple, like I was last night, I logged the first byte of each tuple field in 1.69.0 and 1.71.0+:

```
slot_int, meta_int, account_meta_int, data_int, offset_int, hash_int
1.69.0: 112 160 96 40 104 37
1.71.0: 104 96 160 40 48 37
```
Note the underlying data array bytes are in reverse order: [160..1].

So in memory the layout order is:
```
1.69.0: [meta, slot, offset, account_meta, data, hash]
1.71.0: [account_meta, slot, meta, offset, data, hash] 
```
both still take 160 bytes, but just have a different order 🤷‍♂️

File: accounts-db/src/accounts_db.rs
Comment (ilya-bobyr) [inline accounts-db/src/accounts_db.rs:12612]: Why are the hashes different for debug and release builds?
Doesn't it indicate there is still some unresolved instability?

Comment (ryoqun) [reply accounts-db/src/accounts_db.rs:12612]: good point. I'm sure some rustc optimization is causing the difference.
it indicates some unresolved UB but i think this is tolerable and stable as long as this shouldn't change that often after `repr(C)`.

Comment (ilya-bobyr) [reply accounts-db/src/accounts_db.rs:12612]: The issue is indeed that this test assigns a value that is not 0 or 1 into the `bool` fields, which is an undefined behavior.

Here is a change that makes the hashes identical: https://github.com/solana-labs/solana/pull/33125

I still think that it is better to remove `unsafe` and let the compiler make sure that these tricky issues do not happen: https://github.com/solana-labs/solana/pull/33083

**Pull request merged**
